### 1) Default baseline implementation without storing data in vector DB instead used pytorch tensors

In [4]:
import json
import logging
import pandas as pd

from financerag.tasks import FinDER
from financerag.retrieval import DenseRetrieval, SentenceTransformerEncoder
from financerag.tasks.BaseTask import BaseTask  # optional if you prefer static evaluate

logging.basicConfig(level=logging.INFO)

CORPUS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/finder_corpus.jsonl/corpus.jsonl"
QUERY_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/finder_queries.jsonl/queries.jsonl"
QRELS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/FinDER_qrels.tsv"

def load_jsonl(path):
    with open(path, "r") as f:
        for line in f:
            yield json.loads(line)

class LocalFinDER(FinDER):
    def load_data(self):
        # override BaseTask.load_data so it doesn't try HF
        self.queries = {}
        self.corpus = {}

corpus = {
    doc["_id"]: {"title": doc.get("title", ""), "text": doc.get("text", "")}
    for doc in load_jsonl(CORPUS_PATH)
}
queries = {q["_id"]: q["text"] for q in load_jsonl(QUERY_PATH)}

finder_task = LocalFinDER()
finder_task.corpus = corpus
finder_task.queries = queries

df = pd.read_csv(QRELS_PATH, sep="\t")
qrels = df.groupby("query_id").apply(lambda g: dict(zip(g["corpus_id"], g["score"]))).to_dict()

print("Querry Result", qrels)

encoder = SentenceTransformerEncoder(
    model_name_or_path="intfloat/e5-large-v2",
    query_prompt="query: ",
    doc_prompt="passage: ",
)
retriever = DenseRetrieval(model=encoder)

import time

# Measure just the retrieval call time
start = time.perf_counter()
retrieval_result = finder_task.retrieve(retriever=retriever, top_k=100)
retrieval_time_s = time.perf_counter() - start

print(f"Retrieval time (DenseRetrieval): {retrieval_time_s:.2f}s")

ndcg, map_, recall, precision = finder_task.evaluate(
    qrels=qrels,
    results=retrieval_result,
    k_values=[1, 5, 10],
)
print("NDCG:", ndcg)
print("MAP:", map_)
print("Recall:", recall)
print("Precision:", precision)

finder_task.save_results(output_dir="./results")


/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_69233/1032852643.py:37: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  qrels = df.groupby("query_id").apply(lambda g: dict(zip(g["corpus_id"], g["score"]))).to_dict()
INFO:sentence_transformers.SentenceTransformer:Use pytorch device_name: mps
INFO:sentence_transformers.SentenceTransformer:Load pretrained SentenceTransformer: intfloat/e5-large-v2


Querry Result {'q00001': {'MSFT20230014': 1, 'MSFT20230015': 1}, 'q00007': {'MSFT20231529': 1}, 'q00008': {'MSFT20231529': 1}, 'q00010': {'ADBE20231571': 1, 'ADBE20231572': 1, 'ADBE20230728': 1, 'ADBE20231573': 1}, 'q00019': {'CPNG20230732': 1}, 'q00021': {'CPNG20230658': 1}, 'q00022': {'CPNG20230553': 1}, 'q00027': {'LIN20231133': 1}, 'q00028': {'LIN20231195': 1}, 'q00030': {'LIN20230064': 1, 'LIN20230065': 1, 'LIN20230066': 1, 'LIN20230067': 1}, 'q00034': {'LIN20230551': 1}, 'q00039': {'ORCL20230738': 1, 'ORCL20230739': 1}, 'q00042': {'ORCL20230129': 1, 'ORCL20230130': 1, 'ORCL20230131': 1}, 'q00043': {'ORCL20231527': 1, 'ORCL20231529': 1}, 'q00044': {'ORCL20231505': 1}, 'q00048': {'NVDA20231260': 1}, 'q00062': {'PG20230221': 1, 'PG20230805': 1}, 'q00067': {'PG20230429': 1}, 'q00070': {'PG20230438': 1}, 'q00071': {'PG20230440': 1}, 'q00078': {'DAL20230459': 1}, 'q00081': {'DAL20230513': 1}, 'q00092': {'TSLA20230391': 1}, 'q00095': {'TSLA20231453': 1, 'TSLA20231454': 1}, 'q00099': {'N

INFO:financerag.retrieval.dense:Encoding queries...
Batches: 100%|██████████| 4/4 [00:01<00:00,  2.45it/s]
INFO:financerag.retrieval.dense:Sorting corpus by document length...
INFO:financerag.retrieval.dense:Encoding corpus in batches... This may take a while.
INFO:financerag.retrieval.dense:Encoding batch 1/1...
Batches: 100%|██████████| 217/217 [1:40:08<00:00, 27.69s/it]   
INFO:financerag.tasks.BaseTask:For evaluation, we ignore identical query and document ids (default), please explicitly set ``ignore_identical_ids=False`` to ignore this.
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:NDCG@1: 0.3125
INFO:financerag.tasks.BaseTask:NDCG@5: 0.3860
INFO:financerag.tasks.BaseTask:NDCG@10: 0.4281
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:MAP@1: 0.2531
INFO:financerag.tasks.BaseTask:MAP@5: 0.3465
INFO:financerag.tasks.BaseTask:MAP@10: 0.3685
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:Recall@1: 0.2531
INFO:financerag.tasks.BaseTas

Retrieval time (DenseRetrieval): 2748.75s
NDCG: {'NDCG@1': 0.3125, 'NDCG@5': 0.38599, 'NDCG@10': 0.42814}
MAP: {'MAP@1': 0.25312, 'MAP@5': 0.34648, 'MAP@10': 0.36851}
Recall: {'Recall@1': 0.25312, 'Recall@5': 0.45937, 'Recall@10': 0.57422}
Precision: {'P@1': 0.3125, 'P@5': 0.12188, 'P@10': 0.07969}


![image.png](attachment:image.png)

Enbedding Model: https://huggingface.co/intfloat/e5-large-v2

This model has 24 layers and the embedding size is 1024.


- Corpus: 13,867 documents (from finder_corpus.jsonl)
- Queries: 218 queries (from finder_queries.jsonl)
- Encoding: The model encodes all documents in 217 batches for the corpus and 4 batches for queries
- Reranking: 675 batches of query-document pairs for cross-encoder reranking

Batching is used for memory efficiency and computational performance:

- Query Encoding (4 batches): Your 128 queries are encoded in batches to avoid GPU/CPU memory overflow
- Corpus Encoding (217 batches): The ~13,900 documents are too large to encode at once, so they're processed in chunks of 64 documents (default batch size)
- Reranking (675 batches): The cross-encoder processes query-document pairs in batches of 32 to compute relevance scores

### 2) Milvus Benchmark - Test 4 Index Types (FLAT, IVF_FLAT, IVF_SQ8, HNSW)

Each index encodes the corpus separately with e5-large-v2, then builds its own index structure.

In [ ]:
import json
import logging
import pandas as pd
import shutil
import os
import time

from financerag.tasks import FinDER
from financerag.retrieval import MilvusRetrieval, SentenceTransformerEncoder

logging.basicConfig(level=logging.INFO)

CORPUS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/finder_corpus.jsonl/corpus.jsonl"
QUERY_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/finder_queries.jsonl/queries.jsonl"
QRELS_PATH = "/Users/vikashpr/Dev/Python/FinanceRAG/icaif-24-finance-rag-challenge/FinDER_qrels.tsv"

def load_jsonl(path):
    with open(path, "r") as f:
        for line in f:
            yield json.loads(line)

class LocalFinDER(FinDER):
    def load_data(self):
        self.queries = {}
        self.corpus = {}

# Load data
corpus = {
    doc["_id"]: {"title": doc.get("title", ""), "text": doc.get("text", "")}
    for doc in load_jsonl(CORPUS_PATH)
}
queries = {q["_id"]: q["text"] for q in load_jsonl(QUERY_PATH)}

finder_task = LocalFinDER()
finder_task.corpus = corpus
finder_task.queries = queries

df = pd.read_csv(QRELS_PATH, sep="\t")
qrels = df.groupby("query_id").apply(lambda g: dict(zip(g["corpus_id"], g["score"]))).to_dict()

# Encoder
encoder = SentenceTransformerEncoder(
    model_name_or_path="intfloat/e5-large-v2",
    query_prompt="query: ",
    doc_prompt="passage: ",
)

# Clean up old databases
if os.path.exists("./milvus_db"):
    print("🧹 Cleaning up old Milvus databases...")
    shutil.rmtree("./milvus_db")
os.makedirs("./milvus_db", exist_ok=True)

# Index configurations
index_configs = [
    {"name": "FLAT", "index_type": "FLAT", "index_params": {}, "search_params": {}},
    {"name": "IVF_FLAT", "index_type": "IVF_FLAT", "index_params": {"nlist": 128}, "search_params": {"nprobe": 10}},
    {"name": "IVF_SQ8", "index_type": "IVF_SQ8", "index_params": {"nlist": 128}, "search_params": {"nprobe": 10}},
    {"name": "HNSW", "index_type": "HNSW", "index_params": {"M": 16, "efConstruction": 200}, "search_params": {"ef": 100}},
]

print(f"Testing {len(index_configs)} Milvus index configurations")
print(f"Corpus: {len(corpus)} documents")
print(f"Queries: {len(queries)}")
print(f"Expected time: ~20-40 minutes\n")

benchmark_results = []

for idx, config in enumerate(index_configs, 1):
    index_name = config["name"]
    print(f"\n{'='*60}")
    print(f"[{idx}/{len(index_configs)}] Testing: {index_name}")
    print(f"{'='*60}")
    
    db_path = f"./milvus_db/{index_name.lower()}.db"
    
    try:
        # Create Milvus retriever with this index type
        retriever = MilvusRetrieval(
            model=encoder,
            collection_name=f"corpus_{index_name.lower()}",
            uri=db_path,
            batch_size=64,
            embedding_dim=1024,
            index_type=config["index_type"],
            index_params=config["index_params"],
            search_params=config["search_params"]
        )
        
        # First run: encode corpus + build index, measure retrieval call time
        print(f"📊 Encoding corpus & building {index_name} index...")
        start = time.time()
        retrieval_result = finder_task.retrieve(retriever=retriever, top_k=100)
        index_time = time.time() - start
        if hasattr(retriever, "get_timing_metrics"):
            t = retriever.get_timing_metrics()
            print(f"   Retrieval time (with indexing): {t.get('retrieval_time', index_time):.2f}s | avg/query: {t.get('avg_query_time', 0)*1000:.2f}ms")
        else:
            print(f"   Retrieval time (with indexing): {index_time:.2f}s")
        
        # Second run: search only (corpus already indexed)
        print(f"🔍 Performing search with {index_name}...")
        start = time.time()
        retrieval_result = finder_task.retrieve(retriever=retriever, top_k=100)
        search_time = time.time() - start
        if hasattr(retriever, "get_timing_metrics"):
            t = retriever.get_timing_metrics()
            print(f"   Retrieval time (search only): {t.get('retrieval_time', search_time):.2f}s | avg/query: {t.get('avg_query_time', 0)*1000:.2f}ms")
        else:
            print(f"   Retrieval time (search only): {search_time:.2f}s")
        
        # Evaluate
        print(f"📈 Evaluating {index_name}...")
        ndcg, map_, recall, precision = finder_task.evaluate(
            qrels=qrels, results=retrieval_result, k_values=[1, 5, 10]
        )
        
        result = {
            "index_type": index_name,
            "index_time_sec": round(index_time, 2),
            "search_time_sec": round(search_time, 2),
            "ndcg@10": round(ndcg["NDCG@10"], 4),
        }
        benchmark_results.append(result)
        
        print(f"✅ {index_name} Complete:")
        print(f"   Index time: {result['index_time_sec']}s")
        print(f"   Search time: {result['search_time_sec']}s")
        print(f"   NDCG@10: {result['ndcg@10']}")
        
    except Exception as e:
        print(f"❌ ERROR with {index_name}: {e}")
        benchmark_results.append({
            "index_type": index_name,
            "index_time_sec": -1,
            "search_time_sec": -1,
            "ndcg@1": -1,
            "ndcg@5": -1,
            "ndcg@10": -1,
            "map@10": -1,
            "recall@10": -1,
        })
        continue

print(f"\n{'='*60}")
print("✅ Benchmark Complete!")
print(f"{'='*60}")

/var/folders/_d/jnx4p5xs4854n467vl8chy_m0000gn/T/ipykernel_47390/192954063.py:39: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  qrels = df.groupby("query_id").apply(lambda g: dict(zip(g["corpus_id"], g["score"]))).to_dict()
INFO:sentence_transformers.SentenceTransformer:Use pytorch device_name: mps
INFO:sentence_transformers.SentenceTransformer:Load pretrained SentenceTransformer: intfloat/e5-large-v2
INFO:financerag.retrieval.milvus_retrieval:Created Milvus directory: milvus_db


🧹 Cleaning up old Milvus databases...
Testing 4 Milvus index configurations
Corpus: 13863 documents
Queries: 216
Expected time: ~20-40 minutes


[1/4] Testing: FLAT


huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
INFO:financerag.retrieval.milvus_retrieval:Initialized Milvus client with URI: ./milvus_db/flat.db, Index Type: FLAT
INFO:financerag.retrieval.milvus_retrieval:Starting Milvus-based retrieval...
INFO:financerag.retrieval.milvus_retrieval:Creating new collection 'corpus_flat' with index type: FLAT...


📊 Encoding corpus & building FLAT index...


INFO:financerag.retrieval.milvus_retrieval:Encoding 13863 documents...
Batches: 100%|██████████| 1/1 [00:34<00:00, 34.62s/it]
INFO:financerag.retrieval.milvus_retrieval:Encoded 8000/13863 documents
Batches: 100%|██████████| 1/1 [00:21<00:00, 21.33s/it]
INFO:financerag.retrieval.milvus_retrieval:Encoded 13863/13863 documents
INFO:financerag.retrieval.milvus_retrieval:Adding documents to Milvus...
INFO:financerag.retrieval.milvus_retrieval:Added 1000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 2000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 3000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 4000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 5000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 6000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 7000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieva

🔍 Performing search with FLAT...


Batches: 100%|██████████| 4/4 [00:27<00:00,  6.99s/it]
INFO:financerag.retrieval.milvus_retrieval:Retrieval complete for 216 queries.
INFO:financerag.tasks.BaseTask:For evaluation, we ignore identical query and document ids (default), please explicitly set ``ignore_identical_ids=False`` to ignore this.
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:NDCG@1: 0.3125
INFO:financerag.tasks.BaseTask:NDCG@5: 0.3860
INFO:financerag.tasks.BaseTask:NDCG@10: 0.4281
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:MAP@1: 0.2531
INFO:financerag.tasks.BaseTask:MAP@5: 0.3465
INFO:financerag.tasks.BaseTask:MAP@10: 0.3685
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:Recall@1: 0.2531
INFO:financerag.tasks.BaseTask:Recall@5: 0.4594
INFO:financerag.tasks.BaseTask:Recall@10: 0.5742
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:P@1: 0.3125
INFO:financerag.tasks.BaseTask:P@5: 0.1219
INFO:financerag.tasks.BaseTask:P@10: 0.0797
INFO:financera

📈 Evaluating FLAT...
✅ FLAT Complete:
   Index time: 12719.4s
   Search time: 29.2s
   NDCG@10: 0.4281

[2/4] Testing: IVF_FLAT


INFO:financerag.retrieval.milvus_retrieval:Initialized Milvus client with URI: ./milvus_db/ivf_flat.db, Index Type: IVF_FLAT
INFO:financerag.retrieval.milvus_retrieval:Starting Milvus-based retrieval...
INFO:financerag.retrieval.milvus_retrieval:Creating new collection 'corpus_ivf_flat' with index type: IVF_FLAT...


📊 Encoding corpus & building IVF_FLAT index...


INFO:financerag.retrieval.milvus_retrieval:Encoding 13863 documents...
Batches: 100%|██████████| 1/1 [00:33<00:00, 33.80s/it]
INFO:financerag.retrieval.milvus_retrieval:Encoded 8000/13863 documents
Batches: 100%|██████████| 1/1 [00:18<00:00, 18.69s/it]
INFO:financerag.retrieval.milvus_retrieval:Encoded 13863/13863 documents
INFO:financerag.retrieval.milvus_retrieval:Adding documents to Milvus...
INFO:financerag.retrieval.milvus_retrieval:Added 1000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 2000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 3000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 4000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 5000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 6000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieval:Added 7000/13863 documents to Milvus
INFO:financerag.retrieval.milvus_retrieva

🔍 Performing search with IVF_FLAT...


INFO:financerag.retrieval.milvus_retrieval:Collection 'corpus_ivf_flat' already exists with 13863 documents. Skipping indexing.
INFO:financerag.retrieval.milvus_retrieval:Encoding 216 queries...
Batches: 100%|██████████| 4/4 [00:18<00:00,  4.69s/it]
INFO:financerag.retrieval.milvus_retrieval:Retrieval complete for 216 queries.
INFO:financerag.tasks.BaseTask:For evaluation, we ignore identical query and document ids (default), please explicitly set ``ignore_identical_ids=False`` to ignore this.
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:NDCG@1: 0.3125
INFO:financerag.tasks.BaseTask:NDCG@5: 0.3860
INFO:financerag.tasks.BaseTask:NDCG@10: 0.4281
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:MAP@1: 0.2531
INFO:financerag.tasks.BaseTask:MAP@5: 0.3465
INFO:financerag.tasks.BaseTask:MAP@10: 0.3685
INFO:financerag.tasks.BaseTask:

INFO:financerag.tasks.BaseTask:Recall@1: 0.2531
INFO:financerag.tasks.BaseTask:Recall@5: 0.4594
INFO:financerag.tasks.BaseTask:

📈 Evaluating IVF_FLAT...
✅ IVF_FLAT Complete:
   Index time: 10429.04s
   Search time: 20.16s
   NDCG@10: 0.4281

[3/4] Testing: IVF_SQ8


INFO:financerag.retrieval.milvus_retrieval:Initialized Milvus client with URI: ./milvus_db/ivf_sq8.db, Index Type: IVF_SQ8
INFO:financerag.retrieval.milvus_retrieval:Starting Milvus-based retrieval...
INFO:financerag.retrieval.milvus_retrieval:Creating new collection 'corpus_ivf_sq8' with index type: IVF_SQ8...


📊 Encoding corpus & building IVF_SQ8 index...


INFO:financerag.retrieval.milvus_retrieval:Encoding 13863 documents...
Batches:   0%|          | 0/1 [00:00<?, ?it/s]

In [ ]:
# Display results
results_df = pd.DataFrame(benchmark_results)
print("\n" + "="*80)
print("MILVUS INDEX COMPARISON")
print("="*80)
print(results_df.to_string(index=False))

# Save
output_file = "./results/FinDER/milvus_index_benchmark.csv"
os.makedirs(os.path.dirname(output_file), exist_ok=True)
results_df.to_csv(output_file, index=False)
print(f"\n✅ Saved: {output_file}")

# Best results
valid = results_df[results_df['search_time_sec'] > 0]
if len(valid) > 0:
    best_speed = valid.loc[valid['search_time_sec'].idxmin()]
    best_acc = valid.loc[valid['ndcg@10'].idxmax()]
    print(f"\n🚀 Fastest: {best_speed['index_type']} ({best_speed['search_time_sec']}s)")
    print(f"🎯 Best NDCG@10: {best_acc['index_type']} ({best_acc['ndcg@10']})")